python/pyarrow/_flight.pyx

```diff
@@ -736,16 +742,47 @@ cdef class FlightEndpoint(_Weakrefable):
                     CLocation.Parse(tobytes(location)).Value(&c_location))
             self.endpoint.locations.push_back(c_location)
 
+        if expiration_time is not None:
+            self.endpoint.expiration_time = TimePoint_from_ns(
+                expiration_time.cast(timestamp("ns")).value)
```

Comment:
Looking at how we do `optional` elsewhere, how about:
```suggestion
        if expiration_time is not None:
            self.endpoint.expiration_time = <optional[CTimePoint]>(
                TimePoint_from_ns(expiration_time.value.cast(timestamp("ns"))))
```
Or maybe even:
```suggestion
        self.endpoint.expiration_time = <optional[CTimePoint]>nullopt if expiration_time is None else
            <optional[CTimePoint]>(TimePoint_from_ns(expiration_time.value.cast(timestamp("ns"))))
```

Reply:
I tested this out and it doesn't actually fix the problem, as it isn't the handling of the `optional` that's the problem but the `duration` type parameter of the `time_point`.

I was getting confused trying to understand this error until I realised the type annotation for `CFlightEndpoint.expiration_time` is wrong, it is an `optional[CTimePoint]` in `libarrow_flight.pxd`, but in `cpp/src/arrow/flight/types.h`, it's actually a `std::optional<std::chrono::system_clock::time_point>` (see [types.h](https://github.com/apache/arrow/blob/f3dd298bd32f6dc38654680c49b3f1fbf97e3d5f/cpp/src/arrow/flight/types.h#L83)). `CTimePoint` is `std::chrono::time_point<std::chrono::system_clock, std::chrono::nanoseconds>` (from [python/datetime.h](https://github.com/apache/arrow/blob/f3dd298bd32f6dc38654680c49b3f1fbf97e3d5f/python/pyarrow/src/arrow/python/datetime.h#L88)) and always has nanosecond precision, but `std::chrono::system_clock::time_point` has microsecond precision on MacOS.

Reply:
How about something like this change, which fixes the build: https://github.com/adamreeve/arrow/commit/af9a437f7b23dcbca1307fbb3e4b8c9f28161ea4

Reply:
Thanks for tracking this down. Fixing the declaration sounds like the way to go. 

I wonder if you might be able to "trick" Cython with a declaration like

```
CTimePoint CTimePointFromSystemTime"std::chrono::time_point_cast<TimePoint::duration>"(time_point)
```

but if not it's no big deal

Reply:
@adamreeve thanks for pinpointing the root cause for this. I can confirm changing Timestamp in `cpp/src/arrow/flight/types.h` fixes this:
```diff
-using Timestamp = std::chrono::system_clock::time_point;
+using Timestamp = std::chrono::time_point<std::chrono::system_clock, std::chrono::nanoseconds>;
```
b726b9a0259ac4f97f1cc7c76681b9ac7b80a0ce

Reply:
> I wonder if you might be able to "trick" Cython with a declaration like
> 
> ```
> CTimePoint CTimePointFromSystemTime"std::chrono::time_point_cast<TimePoint::duration>"(time_point)
> ```

How would I get this into the code? When I add it to `libarrow_flight.pxd` and use it in `_flight.pyx`:
```
self.endpoint.expiration_time = CTimePointToSystemTime(TimePoint_from_ns(expiration_time.cast(timestamp("ns")).value))
```
I get the usual:
```
pyarrow/_flight.pyx:748:84: Cannot convert 'CTimePoint' to Python object
```

That "trick" looks a lot like what https://github.com/adamreeve/arrow/commit/af9a437f7b23dcbca1307fbb3e4b8c9f28161ea4 does. Shall we go with that?

Reply:
SGTM